File: PHP/50. Pow(x, n).php
<?php

/**
 * @noinspection PhpIllegalPsrClassPathInspection
 * @noinspection PhpMultipleClassDeclarationsInspection
 * @noinspection PhpUnused
 */
class Solution
{
    public function myPow(float $x, int $n): float
    {
        if ($n < 0) {
            $x = 1 / $x;
            $n = -$n;
        }

        return $this->fastPow($x, $n);
    }

    public function fastPow(float $x, int $n): float
    {
        if ($n === 0) {
            return 1.0;
        }

        $half = $this->fastPow($x, (int) ($n / 2));

        return $n % 2 === 0 ? $half * $half : $half * $half * $x;
    }
}

File: C-Sharp/50. Pow(x, n).cs
using System.Diagnostics.CodeAnalysis;

namespace LeetCode.C_Sharp
{
    [SuppressMessage("ReSharper", "UnusedType.Global")]
    [SuppressMessage("ReSharper", "UnusedMember.Global")]
    public partial class Solution
    {
        public static double MyPow(double x, int n)
        {
            if (n >= 0)
            {
                return FastPow(x, n);
            }

            x = 1 / x;
            n = -n;

            return FastPow(x, n);
        }

        private static double FastPow(double x, int n)
        {
            if (n == 0)
                return 1.0;

            double half = FastPow(x, n / 2);

            if (n % 2 == 0)
                return half * half;

            return half * half * x;
        }
    }
}
